test/util/zip.js

```diff
@@ -23,50 +61,42 @@ const zipStreamToFiles = (zipStream, callback) => {
       setTimeout(() => {
         // eslint-disable-next-line no-shadow
         yauzl.open(tmpfile, { autoClose: false }, (err, zipfile) => {
-          // eslint-disable-next-line keyword-spacing
-          if(err) return callback(err);
-
-          const result = { filenames: [] };
-          const entries = [];
-          let completed = 0;
-
-          zipfile.on('entry', (entry) => entries.push(entry));
-          zipfile.on('end', () => {
-            if (entries.length === 0) {
-              callback(null, result);
-              zipfile.close();
-            } else {
-              entries.forEach((entry) => {
-                result.filenames.push(entry.fileName);
-                // eslint-disable-next-line no-shadow
-                zipfile.openReadStream(entry, (err, resultStream) => {
-                  // eslint-disable-next-line keyword-spacing
-                  if(err) return callback(err);
-
-                  // eslint-disable-next-line no-shadow
-                  resultStream.pipe(streamTest.toText((err, contents) => {
-                    // eslint-disable-next-line keyword-spacing
-                    if(err) return callback(err);
-
-                    result[entry.fileName] = contents;
-                    completed += 1;
-                    if (completed === entries.length) {
-                      callback(null, result);
-                      zipfile.close();
-                    }
-                  }));
-                });
-              });
-            }
-          });
+          if (err) return callback(err);
+          processZipFile(zipfile, callback);
         });
       }, 5); // otherwise sometimes the file doesn't fully drain
     });
   });
 };
 
-// eslint-disable-next-line no-confusing-arrow
-const pZipStreamToFiles = (zipStream) => new Promise((resolve, reject) => { zipStreamToFiles(zipStream, (err, result) => err ? reject(err) : resolve(result)); });
+// Work around a subtle bug in streams that is not fully understood.  Probably one of:
+//
+// * the cleanup code in test/util/zip.js, or
+// * an unfulfilled interaction that supertest is expecting when working with streams, or
+// * a bug in supertest's pipe/stream implementation
+//
+// Possibly: https://github.com/ladjs/supertest/issues/487
+const binaryParser = (res, callback) => {
+  res.setEncoding('binary');
+  let data = '';
+  res.on('data', (chunk) => { data += chunk; });
+  res.on('end', () => {
```

Comment:
Is an error possible here / in this function? I don't see any error handling, but maybe errors aren't possible here.

Reply:
I can't find clear docs on this, but I think this is an example of a `superagent` custom parser, mentioned at https://github.com/ladjs/superagent/blob/master/docs/index.md#parsing-response-bodies

The stock parsers are at https://github.com/ladjs/superagent/tree/2fd631ae225f4335fef37e9888925db0ef42d497/src/node/parsers, and don't seem to take an arg for the `end` event.